pkg/sql/catalog/hydratedtables/hydratedcache_test.go

```diff
@@ -0,0 +1,421 @@
+		retrieved, _, err := c.GetHydratedTableDescriptor(ctx, td, res)
+		require.NoError(t, err)
+		assertMetrics(t, m, 0, 2)
+
+		require.NotEqual(t, hydrated, retrieved)
+	})
+	// If one cache retrieval hits an error during a lookup, it should not
+	// propagate back to a concurrent retrieval as it could be due to something
+	// like cancellation.
+	t.Run("errors do not propagate to concurrent calls", func(t *testing.T) {
+		c := NewCache(cluster.MakeTestingClusterSettings())
+		dg := mkDescGetter(descs...)
+		res := &descGetterTypeDescriptorResolver{dg: &dg}
+		calledCh := make(chan chan error, 1)
+		res.called = func(ctx context.Context, id descpb.ID) error {
+			errCh := make(chan error, 1)
+			calledCh <- errCh
+			return <-errCh
+		}
+		td := tableDescUDT.ImmutableCopy().(*tabledesc.Immutable)
+
+		callOneErrCh := make(chan error, 1)
+		go func() {
+			_, _, err := c.GetHydratedTableDescriptor(ctx, td, res)
+			callOneErrCh <- err
+		}()
+
+		call2Res := &descGetterTypeDescriptorResolver{dg: &dg}
+		unblockCallOne := <-calledCh
+		callTwoErrCh := make(chan error, 1)
+		go func() {
+			_, _, err := c.GetHydratedTableDescriptor(ctx, td, call2Res)
+			callTwoErrCh <- err
+		}()
+		unblockCallOne <- context.Canceled
+		require.Equal(t, context.Canceled, <-callOneErrCh)
+		require.NoError(t, <-callTwoErrCh)
+		assertMetrics(t, c.Metrics(), 0, 1)
+	})
+	t.Run("modified table gets rejected", func(t *testing.T) {
+		c := NewCache(cluster.MakeTestingClusterSettings())
+		dg := mkDescGetter(descs...)
+		res := &descGetterTypeDescriptorResolver{dg: &dg}
+		mut := tabledesc.NewExistingMutable(*dg[tableUDTID].(catalog.TableDescriptor).TableDesc())
+		mut.MaybeIncrementVersion()
+		td := mut.ImmutableCopy().(*tabledesc.Immutable)
+		hydrated, canUse, err := c.GetHydratedTableDescriptor(ctx, td, res)
+		require.NoError(t, err)
+		require.False(t, canUse)
+		require.Nil(t, hydrated)
+	})
+	t.Run("modified type does not get cached", func(t *testing.T) {
+		c := NewCache(cluster.MakeTestingClusterSettings())
+		m := c.Metrics()
+
+		dg := mkDescGetter(descs...)
+		res := &descGetterTypeDescriptorResolver{dg: &dg}
+
+		mut := typedesc.NewExistingMutable(*dg[typ1ID].(catalog.TypeDescriptor).TypeDesc())
+		mut.MaybeIncrementVersion()
+		dgWithMut := mkDescGetter(append(descs, mut)...)
+		resWithMut := &descGetterTypeDescriptorResolver{dg: &dgWithMut}
+
+		// Given that there is no cached value for this version, we will not find
+		// check for a cached type and will construct the hydrated type underneath
+		// the cache. We can use this descriptor, however, it will not be stored.
+		//
+		// This behavior is a bit bizarre but exists to not waste the work of
+		// hydrating the descriptor if we've already started to do it.
+		// This case should not meaningfully arise in practice.
+		td := tableDescUDT.ImmutableCopy().(*tabledesc.Immutable)
+		{
+			hydrated, canUse, err := c.GetHydratedTableDescriptor(ctx, td, resWithMut)
```

Comment:
why can we use cache in this case and below? Isn't the type modified

Reply:
Can use does not mean that we definitely stored it in the cache, just that we hydrated it underneath the call.